### PR TITLE
Testing Walkthrough

### PR DESCRIPTION
**Pull request by betrevisan**
The changes proposed in this PR reflect the tasks in the testing walkthrough, which includes coding the query function for a calendar that finds available time slots for a meeting given its duration, attendees, and other scheduled events and writing tests for that function.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,115 @@
...
+        ArrayList<String> attendeesOptionalList = new ArrayList<String>(attendeesOptional);
+        attendeesRequestList.addAll(attendeesOptionalList);
+
+        List<TimeRange> conflicts = new ArrayList<TimeRange>();
+        Iterator<Event> eventsIterator = events.iterator();
+        while (eventsIterator.hasNext()) {
+            Event event = eventsIterator.next();
+
+            // If there is at least one required or optional attendee in the given event, this event should affect the returned time range.
+            if (!Collections.disjoint(attendeesRequestList, event.getAttendees())) {
```
Comment (kashiish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:69]: Didn't know this was a thing, this is so useful!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,115 @@
...
+
+    public List<TimeRange> findAvailableTimeSlot(List<TimeRange> conflicts, long durationRequest) {
+        Collections.sort(conflicts, TimeRange.ORDER_BY_START);
+
+        List<TimeRange> available = new ArrayList<TimeRange>();
+        int start = TimeRange.START_OF_DAY;
+        int end = TimeRange.END_OF_DAY;
+        int first = 1;
+        for (TimeRange when : conflicts) {
+            if (first == 1) {
```
[the remark this answers is not in the document]
Comment (betrevisan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:92]: Just submitted them on a new commit to this PR

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,115 @@
...
+    }
+
+    public List<TimeRange> findAvailableTimeSlot(List<TimeRange> conflicts, long durationRequest) {
+        Collections.sort(conflicts, TimeRange.ORDER_BY_START);
+
+        List<TimeRange> available = new ArrayList<TimeRange>();
+        int start = TimeRange.START_OF_DAY;
+        int end = TimeRange.END_OF_DAY;
+        int first = 1;
+        for (TimeRange when : conflicts) {
```
[the remark this answers is not in the document]
Comment (betrevisan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Sounds good, I submitted it as a new commit to this PR

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,115 @@
...
+
+            return available;
+        } else {
+            // If there are no optional attendees, just consider the required ones.
+            List<TimeRange> available = disconsiderOptionalAttendees(events, attendeesRequest, durationRequest);
+            return available;
+        }
+    }
+
+    public List<TimeRange> considerOptionalAttendees(Collection<Event> events, Collection<String> attendeesRequest, Collection<String> attendeesOptional, long durationRequest) {
```
[the remark this answers is not in the document]
Comment (betrevisan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: That makes sense, I submitted those changes as a new commit to this PR.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -37,89 +37,102 @@
...
         }
 
-        // If there are at least one optional attendee, consider them.
+        // If there is at least one optional attendee, consider them.
         if (!attendeesOptional.isEmpty()) {
-            List<TimeRange> available = considerOptionalAttendees(events, attendeesRequest, attendeesOptional, durationRequest);
+            List<TimeRange> available = considerAttendees(events, attendeesRequest, attendeesOptional, durationRequest);
 
             // If there was no available time range considering the optional attendees, try just considering the required attendees (if not empty).
             if (available.isEmpty() && !attendeesRequest.isEmpty()) {
```
Comment (dieseldesai) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: sorry for the churn, just noticed another improvement

Here if available is not empty you can just return it immediately. That way the stuff in the else clause can be unindented and you can dedupe the considerAttendees line

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -37,89 +37,102 @@
...
+        for (TimeRange conflict : conflicts) {
+            // If this is the first conflict seen, then set the start and end variables as the start and end of that conflict.
             if (first == 1) {
                 first = 0;
-                start = when.start();
-                end = when.end();
+                start = conflict.start();
+                end = conflict.end();
+
+                // If there is enough time for the requested event to happen before the first conflic, then add the time range from the start
```
Comment (dieseldesai) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:97]: conflict

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -37,89 +37,102 @@
...
     }
 
+    // Returns a list of available time ranges for the requested event, given its duration and a list of time ranges for any conflicting events.
     public List<TimeRange> findAvailableTimeSlot(List<TimeRange> conflicts, long durationRequest) {
         Collections.sort(conflicts, TimeRange.ORDER_BY_START);
 
         List<TimeRange> available = new ArrayList<TimeRange>();
         int start = TimeRange.START_OF_DAY;
         int end = TimeRange.END_OF_DAY;
         int first = 1;
```
Comment (dieseldesai) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: it seems like you're using this as a boolean flag? If so, use a boolean type

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (betrevisan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:97]: Sorry, I didn't quite get what the conflict is here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (betrevisan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Just updated that, thanks!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (betrevisan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: No problem at all, thank you so much for the advice on this!

I just addressed this in a new commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -86,11 +83,11 @@
         List<TimeRange> available = new ArrayList<TimeRange>();
         int start = TimeRange.START_OF_DAY;
         int end = TimeRange.END_OF_DAY;
-        int first = 1;
+        boolean first = true;
         for (TimeRange conflict : conflicts) {
             // If this is the first conflict seen, then set the start and end variables as the start and end of that conflict.
-            if (first == 1) {
-                first = 0;
+            if (first == true) {
```
Comment (dieseldesai) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: you can just write `if (first)`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (dieseldesai) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:94]: Oh, i meant that you typoed the word "conflict" in your comment here.

**Pull request merged**